File: internal/ui/schedule.go
package ui

import (
	"fmt"
	"sort"
	"time"

	"github.com/AmanTahiliani/box-box/internal/models"
)

const defaultSessionDuration = 3 * time.Hour

func parseScheduleTime(value string) (time.Time, error) {
	if value == "" {
		return time.Time{}, fmt.Errorf("empty time value")
	}

	t, err := time.Parse(time.RFC3339, value)
	if err == nil {
		return t.Local(), nil
	}

	t, err = time.Parse("2006-01-02", value[:min(len(value), 10)])
	if err != nil {
		return time.Time{}, err
	}
	return t.Local(), nil
}

func meetingStartTime(meeting models.Meeting) (time.Time, error) {
	return parseScheduleTime(meeting.DateStart)
}

func meetingEndTime(meeting models.Meeting) (time.Time, error) {
	if meeting.DateEnd != "" {
		if end, err := parseScheduleTime(meeting.DateEnd); err == nil {
			return end, nil
		}
	}

	start, err := meetingStartTime(meeting)
	if err != nil {
		return time.Time{}, err
	}
	return start.Add(72 * time.Hour), nil
}

func sessionStartTime(session models.Session) (time.Time, error) {
	return parseScheduleTime(session.DateStart)
}

func sessionEndTime(session models.Session) (time.Time, error) {
	if session.DateEnd != "" {
		if end, err := parseScheduleTime(session.DateEnd); err == nil {
			return end, nil
		}
	}

	start, err := sessionStartTime(session)
	if err != nil {
		return time.Time{}, err
	}
	return start.Add(defaultSessionDuration), nil
}

func sortSessionsByStart(sessions []models.Session) {
	sort.Slice(sessions, func(i, j int) bool {
		left, errLeft := sessionStartTime(sessions[i])
		right, errRight := sessionStartTime(sessions[j])
		if errLeft != nil || errRight != nil {
			return sessions[i].DateStart < sessions[j].DateStart
		}
		return left.Before(right)
	})
}

func currentMeeting(meetings []models.Meeting, now time.Time) *models.Meeting {
	var selected *models.Meeting
	var latest time.Time

	for i := range meetings {
		start, err := meetingStartTime(meetings[i])
		if err != nil || start.After(now) {
			continue
		}
		end, err := meetingEndTime(meetings[i])
		if err != nil || now.After(end.Add(24*time.Hour)) {
			continue
		}
		if selected == nil || start.After(latest) {
			selected = &meetings[i]
			latest = start
		}
	}

	return selected
}

func nextUpcomingMeeting(meetings []models.Meeting, now time.Time) *models.Meeting {
	for i := range meetings {
		start, err := meetingStartTime(meetings[i])
		if err != nil {
			continue
		}
		if start.After(now) {
			return &meetings[i]
		}
	}
	return nil
}

func currentAndNextSession(sessions []models.Session, now time.Time) (*models.Session, *models.Session) {
	for i := range sessions {
		start, err := sessionStartTime(sessions[i])
		if err != nil {
			continue
		}
		end, err := sessionEndTime(sessions[i])
		if err != nil {
			continue
		}

		if (now.Equal(start) || now.After(start)) && now.Before(end) {
			return &sessions[i], nil
		}
		if now.Before(start) {
			return nil, &sessions[i]
		}
	}

	return nil, nil
}

func meetingHasStarted(meeting models.Meeting, now time.Time) bool {
	start, err := meetingStartTime(meeting)
	if err != nil {
		return false
	}
	return !now.Before(start)
}

File: frontend/src/lib/schedule.ts
import type { Meeting, Session } from '../types'

const DEFAULT_SESSION_DURATION_MS = 3 * 60 * 60 * 1000

export function parseScheduleTime(value: string): Date | null {
  if (!value) return null
  const parsed = Date.parse(value)
  if (!Number.isNaN(parsed)) return new Date(parsed)
  const dateOnly = Date.parse(value.slice(0, 10))
  return Number.isNaN(dateOnly) ? null : new Date(dateOnly)
}

export function meetingStartTime(meeting: Meeting): Date | null {
  return parseScheduleTime(meeting.date_start)
}

export function meetingEndTime(meeting: Meeting): Date | null {
  const end = parseScheduleTime(meeting.date_end)
  if (end) return end
  const start = meetingStartTime(meeting)
  return start ? new Date(start.getTime() + 72 * 60 * 60 * 1000) : null
}

export function sessionStartTime(session: Session): Date | null {
  return parseScheduleTime(session.date_start)
}

export function sessionEndTime(session: Session): Date | null {
  const end = parseScheduleTime(session.date_end)
  if (end) return end
  const start = sessionStartTime(session)
  return start ? new Date(start.getTime() + DEFAULT_SESSION_DURATION_MS) : null
}

export function sortSessionsByStart(sessions: Session[]): Session[] {
  return [...sessions].sort((a, b) => {
    const left = sessionStartTime(a)?.getTime() ?? 0
    const right = sessionStartTime(b)?.getTime() ?? 0
    if (left !== right) return left - right
    return a.date_start.localeCompare(b.date_start)
  })
}

export function currentMeeting(meetings: Meeting[], now: Date): Meeting | null {
  let selected: Meeting | null = null
  let latest = 0

  for (const meeting of meetings) {
    const start = meetingStartTime(meeting)
    if (!start || start > now) continue
    const end = meetingEndTime(meeting)
    if (!end || now > new Date(end.getTime() + 24 * 60 * 60 * 1000)) continue
    const startMs = start.getTime()
    if (!selected || startMs > latest) {
      selected = meeting
      latest = startMs
    }
  }

  return selected
}

export function nextUpcomingMeeting(meetings: Meeting[], now: Date): Meeting | null {
  for (const meeting of meetings) {
    const start = meetingStartTime(meeting)
    if (start && start > now) return meeting
  }
  return null
}

export function mostRecentPastMeeting(meetings: Meeting[], now: Date): Meeting | null {
  let selected: Meeting | null = null
  let latest = 0

  for (const meeting of meetings) {
    const start = meetingStartTime(meeting)
    if (!start || start > now) continue
    const startMs = start.getTime()
    if (!selected || startMs > latest) {
      selected = meeting
      latest = startMs
    }
  }

  return selected
}

export function pickFocusMeeting(meetings: Meeting[], now: Date): Meeting | null {
  return (
    currentMeeting(meetings, now) ??
    nextUpcomingMeeting(meetings, now) ??
    mostRecentPastMeeting(meetings, now) ??
    meetings[0] ??
    null
  )
}

export function meetingHasStarted(meeting: Meeting, now: Date): boolean {
  const start = meetingStartTime(meeting)
  return start != null && now >= start
}

export function currentAndNextSession(
  sessions: Session[],
  now: Date,
): { current: Session | null; next: Session | null } {
  const sorted = sortSessionsByStart(sessions)

  for (const session of sorted) {
    const start = sessionStartTime(session)
    const end = sessionEndTime(session)
    if (!start || !end) continue

    if (now >= start && now < end) {
      return { current: session, next: null }
    }
    if (now < start) {
      return { current: null, next: session }
    }
  }

  return { current: null, next: null }
}

export function formatCountdown(target: Date, now: Date): string {
  const diffMs = Math.max(0, target.getTime() - now.getTime())
  const totalSeconds = Math.floor(diffMs / 1000)
  const days = Math.floor(totalSeconds / 86400)
  const hours = Math.floor((totalSeconds % 86400) / 3600)
  const mins = Math.floor((totalSeconds % 3600) / 60)
  const secs = totalSeconds % 60
  return `${days}d ${String(hours).padStart(2, '0')}h ${String(mins).padStart(2, '0')}m ${String(secs).padStart(2, '0')}s`
}

export function formatSessionScheduleTime(value: string): string {
  const date = parseScheduleTime(value)
  if (!date) return '—'
  return date.toLocaleString('en-GB', {
    weekday: 'short',
    day: '2-digit',
    month: 'short',
    hour: '2-digit',
    minute: '2-digit',
    hour12: false,
  })
}

export type FocusMeetingKind = 'current' | 'next' | 'recent' | 'fallback'

export function focusMeetingKind(meeting: Meeting, now: Date): FocusMeetingKind {
  if (currentMeeting([meeting], now)) return 'current'
  if (nextUpcomingMeeting([meeting], now)) return 'next'
  if (mostRecentPastMeeting([meeting], now)) return 'recent'
  return 'fallback'
}

export function focusMeetingLabel(kind: FocusMeetingKind): string {
  switch (kind) {
    case 'current':
      return 'Current Weekend'
    case 'next':
      return 'Next Weekend'
    case 'recent':
      return 'Recent Local Weekend'
    default:
      return 'Weekend'
  }
}
